In [1]:
pip install numpy

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.3.1 -> 25.0.1
[notice] To update, run: C:\Users\LAB4-StudentPC34\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


In [2]:
pip install seaborn

Defaulting to user installation because normal site-packages is not writeableNote: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.3.1 -> 25.0.1
[notice] To update, run: C:\Users\LAB4-StudentPC34\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


In [3]:
pip install matplotlib

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.3.1 -> 25.0.1
[notice] To update, run: C:\Users\LAB4-StudentPC34\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


In [4]:
pip install pandas

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.3.1 -> 25.0.1
[notice] To update, run: C:\Users\LAB4-StudentPC34\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


In [5]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import warnings

In [6]:
accidents = pd.read_csv('datasets\\accident_data.csv')

In [7]:
accidents.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Index                    660679 non-null  object 
 1   Accident_Severity        660679 non-null  object 
 2   Accident Date            660679 non-null  object 
 3   Latitude                 660654 non-null  float64
 4   Light_Conditions         660679 non-null  object 
 5   District Area            660679 non-null  object 
 6   Longitude                660653 non-null  float64
 7   Number_of_Casualties     660679 non-null  int64  
 8   Number_of_Vehicles       660679 non-null  int64  
 9   Road_Surface_Conditions  659953 non-null  object 
 10  Road_Type                656159 non-null  object 
 11  Urban_or_Rural_Area      660664 non-null  object 
 12  Weather_Conditions       646551 non-null  object 
 13  Vehicle_Type             660679 non-null  object 
dtypes: f

In [8]:
accidents.describe(include='all')

Index Accident_Severity Accident Date       Latitude  \
count     660679            660679        660679  660654.000000   
unique    421020                 3          1461            NaN   
top     2.01E+12            Slight    30-11-2019            NaN   
freq      239478            563801           704            NaN   
mean         NaN               NaN           NaN      52.553866   
std          NaN               NaN           NaN       1.406922   
min          NaN               NaN           NaN      49.914430   
25%          NaN               NaN           NaN      51.490690   
50%          NaN               NaN           NaN      52.315641   
75%          NaN               NaN           NaN      53.453452   
max          NaN               NaN           NaN      60.757544   

       Light_Conditions District Area      Longitude  Number_of_Casualties  \
count            660679        660679  660653.000000         660679.000000   
unique                5           422            NaN                   NaN   
top            Daylight    Birmingham            NaN                   NaN   
freq             484880         13491            NaN                   NaN   
mean                NaN           NaN      -1.431210              1.357040   
std                 NaN           NaN       1.383330              0.824847   
min                 NaN           NaN      -7.516225              1.000000   
25%                 NaN           NaN      -2.332291              1.000000   
50%                 NaN           NaN      -1.411667              1.000000   
75%                 NaN           NaN      -0.232869              1.000000   
max                 NaN           NaN       1.762010             68.000000   

        Number_of_Vehicles Road_Surface_Conditions           Road_Type  \
count        660679.000000                  659953              656159   
unique                 NaN                       5                   5   
top                    NaN                     Dry  Single carriageway   
freq                   NaN                  447821              492143   
mean              1.831255                     NaN                 NaN   
std               0.715269                     NaN                 NaN   
min               1.000000                     NaN                 NaN   
25%               1.000000                     NaN                 NaN   
50%               2.000000                     NaN                 NaN   
75%               2.000000                     NaN                 NaN   
max              32.000000                     NaN                 NaN   

       Urban_or_Rural_Area  Weather_Conditions Vehicle_Type  
count               660664              646551       660679  
unique                   3                   8           16  
top                  Urban  Fine no high winds          Car  
freq                421663              520885       497992  
mean                   NaN                 NaN          NaN  
std                    NaN                 NaN          NaN  
min                    NaN                 NaN          NaN  
25%                    NaN                 NaN          NaN  
50%                    NaN                 NaN          NaN  
75%                    NaN                 NaN          NaN  
max                    NaN                 NaN          NaN

In [9]:
accidents.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [10]:
accidents['Latitude'] = accidents['Latitude'].fillna(accidents['Latitude'].mode()[0])

In [11]:
accidents.isnull().sum()


Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                       0
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [12]:
accidents['Longitude'] = accidents['Longitude'].fillna(accidents['Longitude'].mode()[0])

In [13]:
accidents.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                       0
Light_Conditions               0
District Area                  0
Longitude                      0
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [14]:
accidents['Road_Surface_Conditions'] = accidents['Road_Surface_Conditions'].fillna(accidents['Road_Surface_Conditions'].mode()[0])

In [15]:
accidents.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                       0
Light_Conditions               0
District Area                  0
Longitude                      0
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions        0
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [16]:
accidents['Urban_or_Rural_Area'] = accidents['Urban_or_Rural_Area'].fillna(accidents['Urban_or_Rural_Area'].mode()[0])

In [17]:
accidents.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                       0
Light_Conditions               0
District Area                  0
Longitude                      0
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions        0
Road_Type                   4520
Urban_or_Rural_Area            0
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [18]:
accidents['Road_Type'] = accidents['Road_Type'].fillna("unknown")

In [19]:
accidents.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                       0
Light_Conditions               0
District Area                  0
Longitude                      0
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions        0
Road_Type                      0
Urban_or_Rural_Area            0
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [20]:
accidents['Weather_Conditions'] = accidents['Weather_Conditions'].fillna("unaccounted")

In [21]:
accidents.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
dtype: int64

<h1>filling up null values

In [22]:
accidents.dtypes

Index                       object
Accident_Severity           object
Accident Date               object
Latitude                   float64
Light_Conditions            object
District Area               object
Longitude                  float64
Number_of_Casualties         int64
Number_of_Vehicles           int64
Road_Surface_Conditions     object
Road_Type                   object
Urban_or_Rural_Area         object
Weather_Conditions          object
Vehicle_Type                object
dtype: object

In [23]:
accidents['Accident Date'] = pd.to_datetime(accidents['Accident Date'], dayfirst = True, errors = 'coerce')
accidents["Year"] = accidents['Accident Date'].dt.year
accidents["Month"] = accidents['Accident Date'].dt.month
accidents["Day"] = accidents['Accident Date'].dt.day
accidents["DayOfWeek"] = accidents['Accident Date'].dt.dayofweek

In [24]:
accidents

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious    2019-06-05  51.506187   
1       200701BS65737           Serious    2019-07-02  51.495029   
2       200701BS66127           Serious           NaT  51.517715   
3       200701BS66128           Serious           NaT  51.495478   
4       200701BS66837            Slight    2019-09-03  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight           NaT  57.374005   
660675  201091NM01881            Slight           NaT  57.232273   
660676  201091NM01935            Slight           NaT  57.585044   
660677  201091NM01964           Serious           NaT  57.214898   
660678  201091NM02142           Serious           NaT  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                  unknown               Urban            unaccounted   
3       Single carriageway               Urban     Fine no high winds   
4                  unknown               Urban            unaccounted   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type    Year  Month  Day  DayOfWeek  
0                                        Car  2019.0    6.0  5.0        2.0  
1                                        Car  2019.0    7.0  2.0        1.0  
2                      Taxi/Private hire car     NaN    NaN  NaN        NaN  
3       Bus or coach (17 or more pass seats)     NaN    NaN  NaN        NaN  
4                              Other vehicle  2019.0    9.0  3.0        1.0  
...                                      ...     ...

In [25]:
accidents.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 18 columns):
 #   Column                   Non-Null Count   Dtype         
---  ------                   --------------   -----         
 0   Index                    660679 non-null  object        
 1   Accident_Severity        660679 non-null  object        
 2   Accident Date            265007 non-null  datetime64[ns]
 3   Latitude                 660679 non-null  float64       
 4   Light_Conditions         660679 non-null  object        
 5   District Area            660679 non-null  object        
 6   Longitude                660679 non-null  float64       
 7   Number_of_Casualties     660679 non-null  int64         
 8   Number_of_Vehicles       660679 non-null  int64         
 9   Road_Surface_Conditions  660679 non-null  object        
 10  Road_Type                660679 non-null  object        
 11  Urban_or_Rural_Area      660679 non-null  object        
 12  Weather_Conditio

In [26]:
accidents.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 18 columns):
 #   Column                   Non-Null Count   Dtype         
---  ------                   --------------   -----         
 0   Index                    660679 non-null  object        
 1   Accident_Severity        660679 non-null  object        
 2   Accident Date            265007 non-null  datetime64[ns]
 3   Latitude                 660679 non-null  float64       
 4   Light_Conditions         660679 non-null  object        
 5   District Area            660679 non-null  object        
 6   Longitude                660679 non-null  float64       
 7   Number_of_Casualties     660679 non-null  int64         
 8   Number_of_Vehicles       660679 non-null  int64         
 9   Road_Surface_Conditions  660679 non-null  object        
 10  Road_Type                660679 non-null  object        
 11  Urban_or_Rural_Area      660679 non-null  object        
 12  Weather_Conditio

In [27]:
accidents['Index'] = accidents['Index'].astype('category')

In [28]:
accidents['Accident_Severity'] = accidents['Accident_Severity'].astype('category')

In [29]:
accidents['Light_Conditions'] = accidents['Light_Conditions'].astype('category')

In [30]:
accidents['District Area'] = accidents['District Area'].astype('category')

In [31]:
accidents['Road_Surface_Conditions'] = accidents['Road_Surface_Conditions'].astype('category')

In [32]:
accidents['Road_Type'] = accidents['Road_Type'].astype('category')

In [33]:
accidents['Urban_or_Rural_Area'] = accidents['Urban_or_Rural_Area'].astype('category')

In [34]:
accidents['Weather_Conditions'] = accidents['Weather_Conditions'].astype('category')

In [35]:
accidents['Vehicle_Type'] = accidents['Vehicle_Type'].astype('category')

In [36]:
accidents.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 18 columns):
 #   Column                   Non-Null Count   Dtype         
---  ------                   --------------   -----         
 0   Index                    660679 non-null  category      
 1   Accident_Severity        660679 non-null  category      
 2   Accident Date            265007 non-null  datetime64[ns]
 3   Latitude                 660679 non-null  float64       
 4   Light_Conditions         660679 non-null  category      
 5   District Area            660679 non-null  category      
 6   Longitude                660679 non-null  float64       
 7   Number_of_Casualties     660679 non-null  int64         
 8   Number_of_Vehicles       660679 non-null  int64         
 9   Road_Surface_Conditions  660679 non-null  category      
 10  Road_Type                660679 non-null  category      
 11  Urban_or_Rural_Area      660679 non-null  category      
 12  Weather_Conditio

In [37]:
accidents.isnull().sum()

Index                           0
Accident_Severity               0
Accident Date              395672
Latitude                        0
Light_Conditions                0
District Area                   0
Longitude                       0
Number_of_Casualties            0
Number_of_Vehicles              0
Road_Surface_Conditions         0
Road_Type                       0
Urban_or_Rural_Area             0
Weather_Conditions              0
Vehicle_Type                    0
Year                       395672
Month                      395672
Day                        395672
DayOfWeek                  395672
dtype: int64

In [38]:
accidents['Accident Date'] = accidents['Accident Date'].fillna("Accident Date")
accidents['Year'] = accidents['Year'].fillna("Year")
accidents['Month'] = accidents['Month'].fillna("Month")
accidents['Day'] = accidents['Day'].fillna("Day")
accidents['DayOfWeek'] = accidents['DayOfWeek'].fillna("DayOfWeek")



In [39]:
accidents.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
Year                       0
Month                      0
Day                        0
DayOfWeek                  0
dtype: int64

In [40]:
#accident_per_year = accidents['Year'].value_counts().sort_index()
#accident_per_month = accidents['Month'].value_counts().sort_index()

#accident_per_year.plot(kind='bar')
#plt.title('Accident per year')
#plt.xlabel('Year')
#plt.ylabel('Number of Accidents')
#plt.show()


month_names = {
    1: 'January', 2: 'February', 3: 'March', 4: 'April', 5: 'May', 6: 'June', 7: 'July' , 8: 'August', 9: 'September', 10:'October', 11: 'November', 12:'December'}

<h1>LINE PLOT</h1>

In [41]:
year_area = accidents.groupby(['District Area', 'Year']).size().reset_index(name = 'Accident_Count')

easy_df = year_area.pivot(index='Year', columns='District Area' , values='Accident_Count')

plt.plot(easy_df.index, easy_df['Aberdeen City'], label='Aberdeen City', marker='o')


year_area.plot(kind='bar')

plt.legend
plt.show()

C:\Users\LAB4-StudentPC34\AppData\Local\Temp\ipykernel_18068\3466995792.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  year_area = accidents.groupby(['District Area', 'Year']).size().reset_index(name = 'Accident_Count')


TypeError: '<' not supported between instances of 'str' and 'float'

In [ ]:
cas_veh = accidents.groupby(['Accident_Severity', 'Vehicle_Type']).size().unstack()

sns.heatmap(cas_veh , cmap = 'Wistia')
plt.show()

<h1>INSIGHT 1. number or accidents in year 2019</h1>

In [ ]:
df_2019 = accidents[accidents['Year'] == 2019]


In [ ]:
df_2019

In [ ]:
accidents['Year'] == 2019

In [ ]:
accident_per_year = accidents['Year'].value_counts()


accident_per_year.plot(kind='bar')
plt.title('Accident per year')
plt.xlabel('Year')
plt.ylabel('Number of Accidents')
plt.show()


In [ ]:
plt.figure(figsize=(10, 5))
plt.xticks(ticks=range(12), labels=['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'])
plt.xlabel("Month")
plt.ylabel("Number of Accidents")
plt.title("Number of Accidents per Month in 2019")
plt.show()

In [ ]:
df_2019 = accidents[accidents['Year'] == 2019]
accident_per_month_2019 = df_2019['Month'].value_counts().sort_index()

# Plot accidents per month in 2019
plt.figure(figsize=(10, 5))
plt.bar(accident_per_month_2019.index - 1, accident_per_month_2019.values)  # Adjusting for months 1-12
plt.xticks(ticks=range(12), labels=['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'])
plt.xlabel("Month")
plt.ylabel("Number of Accidents")
plt.title("Number of Accidents per Month in 2019")
plt.show()

# # Grouping accidents by 'District Area' and 'Year'
# year_area = accidents.groupby(['District Area', 'Year']).size().reset_index(name='Accident_Count')

# # Pivot the data to create a table where 'Year' is the index and 'District Area' is the columns
easy_df = year_area.pivot(index='Year', columns='District Area', values='Accident_Count')

# # Plot accidents by District Area over the years
plt.figure(figsize=(10, 6))
for district in easy_df.columns:
    plt.plot(easy_df.index, easy_df['Aberdeen City'], label=district, marker='o')

plt.title("Accidents by District Area and Year")
plt.xlabel("Year")
plt.ylabel("Number of Accidents")
plt.legend()
plt.show()

# Optionally, you can create a bar plot to show the same data
year_area_pivot = year_area.pivot_table


<h1>INSIGHT 2. Accidents with rainy conditions </h1>

In [ ]:
df_rainy = accidents[accidents['Weather_Conditions'] == 'Rain']


In [ ]:
df_rainy

In [ ]:
rainy_accidents = accidents[accidents['Weather_Conditions'] == 'Rain']
plt.title('Accidents During Rainy Conditions')
plt.xlabel('Weather Condition')
plt.ylabel('Number of Accidents')
plt.show()



In [ ]:
df_daylight = accidents[accidents['Light_Conditions'] == 'Daylight']


In [ ]:
df_daylight

In [ ]:
# df_daylight = accidents['df_daylight'].value_counts()
# df_daylight.plot(kind='bar')
# plt.title('Accidents in daylight')
# plt.xlabel('Light_Conditions')
# plt.ylabel('Number_of_Casualties')
# plt.show()

In [ ]:
df_dry_roads = accidents[accidents['Road_Surface_Conditions'] == 'Dry']


In [ ]:
df_dry_roads

In [ ]:
df_dry_roads = accidents.groupby('Road_Surface_Conditions')['Number_of_Casualties'].sum()
df_dry_roads.plot(kind='bar')
plt.title('Accidents in dry roads')
plt.xlabel('Road_Surface_Conditions')
plt.ylabel('Number_of_Casualties')
plt.show()

<h1>INSIGHT 3. Most common vehicle type in accidents?</h3>

<h5>Cars are involved in the majority of accidents.</h5>

In [ ]:
vehicle_counts = accidents["Vehicle_Type"].value_counts()

In [ ]:
vehicle_counts

In [ ]:
vehicle = vehicle_counts.idxmax()

In [ ]:
vehicle

In [ ]:
vehicle_counts = accidents["Vehicle_Type"].value_counts().head(10)
plt.figure(figsize=(10, 5))
sns.barplot(y=vehicle_counts.index, x=vehicle_counts.values)
plt.xlabel("Number of Accidents")
plt.ylabel("Vehicle Type")
plt.title("Most Common Vehicle Types Involved in Accidents")
plt.show()



<h1>INSIGHT 4.
Most common accident severity type in accidents?

<h5>The most common accident severity level is Slight.</h3>

In [ ]:
severity_counts = accidents["Accident_Severity"].value_counts()

In [ ]:
severity_counts

In [ ]:
severity = severity_counts.idxmax()

In [ ]:
severity

In [ ]:
severity_counts = accidents["Accident_Severity"].value_counts()
sns.barplot(x=severity_counts.index, y=severity_counts.values)
plt.xlabel("Accident Severity")
plt.ylabel("Number of Accidents")
plt.title("Most Common Accident Severity Type")
plt.show()


In [ ]:
severity_counts = accidents["Accident_Severity"].value_counts()


sns.scatterplot(x=severity_counts.index, y=severity_counts.values, s=100, color='blue', marker='o')

plt.xlabel("Accident Severity")
plt.ylabel("Number of Accidents")
plt.title("Most Common Accident Severity Type")
plt.show()


<h1>INSIGHT 5. what district area has the highest accidents?

<h5>Birmingham records the highest number of accidents.</h3>

In [ ]:
district_counts = accidents["District Area"].value_counts()

In [ ]:
district_counts

In [ ]:
district = district_counts.idxmax()

In [ ]:
district

In [ ]:
district_counts = accidents["District Area"].value_counts()

plt.figure(figsize=(20, 5))
sns.scatterplot(x=district_counts.index, y=district_counts.values, s=100, color='green', marker='o')

plt.xlabel("District Area")
plt.ylabel("Number of Accidents")
plt.title("Accidents per District Area")
plt.show()


<h1>INSIGHT 6. on what type of road has the highest number of accidents occur?

<h5>A significant number of accidents take place on Single carriageway roads.</h5>

In [ ]:
road_type_counts = accidents["Road_Type"].value_counts()

In [ ]:
road_type_counts

In [ ]:
road_type = road_type_counts.idxmax()

In [ ]:
road_type

In [ ]:
road_type_counts = accidents["Road_Type"].value_counts()

plt.figure(figsize=(10, 5))
sns.lineplot(x=road_type_counts.index, y=road_type_counts.values, marker='o', color='purple')

plt.xlabel("Road Type")
plt.ylabel("Number of Accidents")
plt.title("Accidents by Road Type")
plt.xticks(rotation=45)  
plt.show()


<h1>INSIGHT 7. most common weather conditions during accidents?</h1>

<h5>The most frequent weather condition during accidents is clear skies with no strong winds.</h5>

In [ ]:
weather_counts = accidents["Weather_Conditions"].value_counts()

In [ ]:
weather_counts

In [ ]:
weather = weather_counts.idxmax()

In [ ]:
weather

In [ ]:
weather_counts = accidents["Weather_Conditions"].value_counts()


weather_counts_df = pd.DataFrame(weather_counts).reset_index()
weather_counts_df.columns = ['Weather_Condition', 'Number_of_Accidents']

heatmap_data = weather_counts_df.set_index('Weather_Condition').T

# Create a heatmap
plt.figure(figsize=(10, 2))
sns.heatmap(heatmap_data, annot=True, cmap='Blues', cbar_kws={'label': 'Number of Accidents'}, fmt='d')

plt.title("Heatmap of Accidents by Weather Conditions")
plt.show()


In [ ]:
# weather = accidents.groupby('Weather_Conditions')['Number_of_Casualties'].sum().reset_index()
# sns.barplot(x=accidents.index, y=accidents.values)
# x = weather['Number_of_Casualties']
# y = weather['Weather_Conditions']
# plt.bar(x,y)
# plt.show()

<h1>INSIGHT 8. most accidents happen in?</h1>

<h5>Accidents are more prevalent in Urban areas.</h5>

In [ ]:
urban_rural = accidents["Urban_or_Rural_Area"].value_counts()

In [ ]:
urban_rural

In [ ]:
urban_or_rural = urban_rural.idxmax()

In [ ]:
urban_or_rural

In [ ]:
plt.figure(figsize=(8, 5))
sns.countplot(data=accidents, x='Urban_or_Rural_Area', palette='dark')
plt.title('Accidents in Urban vs Rural Areas')
plt.xlabel('Area Type')
plt.ylabel('Number of Accidents')
plt.show()

<h1>INSIGHT 9. Average number of casualties per accident</h1>

<h5>average number of casualties per accident is  1.35</h5>

In [ ]:
avg_casualties = accidents["Number_of_Casualties"].mean()

In [ ]:
avg_casualties

In [ ]:
avg_casualties = accidents['Number_of_Casualties'].mean()
plt.figure(figsize=(6, 4))
plt.scatter(['Average Casualties'], [avg_casualties], color='lightcoral')
plt.title('Average Number of Casualties per Accident')
plt.ylabel('Average Casualties')
plt.show()


<h1>INSIGHT 10. most common light condition happens in?</h1>

<h5>The most common lighting condition during accidents is Daylight.</h5>

In [ ]:
light_conditions = accidents["Light_Conditions"].value_counts()

In [ ]:
light_conditions

In [ ]:
light = light_conditions.idxmax()

In [ ]:
light

In [ ]:
plt.figure(figsize=(8, 5))
sns.countplot(data=accidents, x='Light_Conditions', palette='autumn')
plt.title('Number of Accidents by Light Conditions')
plt.xlabel('Light Conditions')
plt.ylabel('Number of Accidents')
plt.xticks(rotation=45)
plt.show()

<h1>INSIGHT 11. highest avg of casualties</h1>

<h5>the highest average casualties is Rural.</h5>

In [ ]:
highest_avg_casualties = accidents.groupby('Urban_or_Rural_Area')['Number_of_Casualties'].mean()


In [ ]:
highest_avg_casualties

In [ ]:
highest = highest_avg_casualties.idxmax()

In [ ]:
highest

In [ ]:
avg_casualties_area = accidents.groupby('Urban_or_Rural_Area')['Number_of_Casualties'].mean().sort_values()
plt.figure(figsize=(8, 5))
avg_casualties_area.plot(kind='bar', color='salmon')
plt.title('Average Casualties by Area Type')
plt.xlabel('Area Type')
plt.ylabel('Average Number of Casualties')
plt.show()

<h1>INSIGHT 12. avg number of vehicles involved in accidents</h1>

<h5>he average number of vehicles involved in an accident is 1.83.</h5>

In [ ]:
vehicles = accidents["Number_of_Vehicles"].mean()

In [ ]:
vehicles

In [ ]:
avg_vehicles = accidents['Number_of_Vehicles'].mean()
plt.figure(figsize=(6, 4))
plt.bar(['Average Vehicles'], [avg_vehicles], color='lightgreen')
plt.title('Average Number of Vehicles Involved in Accidents')
plt.ylabel('Average Vehicles')
plt.show()

<h1>INSIGHTS 13. vehicle type w the least accidents</h1>

<H5>the vehicle with the least accidents is the ridden horse</H5>

In [ ]:
least_vehicle = accidents['Vehicle_Type'].value_counts()

In [ ]:
least_vehicle

In [ ]:
vehicle = vehicle_counts.idxmin()

In [ ]:
vehicle

In [ ]:
least_vehicle_type = accidents['Vehicle_Type'].value_counts().idxmin()
vehicle_counts = accidents['Vehicle_Type'].value_counts()
plt.figure(figsize=(15, 6))
vehicle_counts.plot(kind='bar', color='slateblue')
plt.title('Number of Accidents by Vehicle Type')
plt.xlabel('Vehicle Type')
plt.ylabel('Number of Accidents')
plt.xticks(rotation=45)
plt.show()


<H1>INSIGHT 14. the vehicle type with the highest casualty rate </H1>

In [ ]:
vehicle_heighest_casualties = accidents.groupby("Vehicle_Type")["Number_of_Casualties"].mean().idxmax()

In [ ]:
vehicle_heighest_casualties

In [ ]:
vehicle_highest_casualties = accidents.groupby('Vehicle_Type')['Number_of_Casualties'].mean().sort_values()
plt.figure(figsize=(10, 6))
vehicle_highest_casualties.plot(kind='bar', color='darkred')
plt.title('Average Casualties by Vehicle Type')
plt.xlabel('Vehicle Type')
plt.ylabel('Average Number of Casualties')
plt.xticks(rotation=45)
plt.show()


<h1>INSIGHT 15. region with the highest percentage of severe accidents</h1>

In [ ]:
severe_accidents = accidents[accidents["Accident_Severity"] == "Severe"]

In [ ]:
region = severe_accidents['District Area'].value_counts().idxmax()

In [ ]:
region

In [ ]:
severe_accidents = accidents[accidents['Accident_Severity'] == 'Severe']
region = severe_accidents['District Area'].value_counts()

region.plot(kind='line', color='orange')
plt.title('Regions with Highest Severe Accidents')
plt.xlabel('District Area')
plt.ylabel('Number of Severe Accidents')
plt.show()


<h1>INSIGHT 16. wet and icy roads show higher casualty rates</h1>

In [ ]:
accidents_urban_rural = accidents.groupby(['Urban_or_Rural_Area', 'Accident_Severity']).size().unstack()

In [ ]:
accidents_urban_rural

In [ ]:
accidents_urban_rural = accidents.groupby(['Urban_or_Rural_Area', 'Accident_Severity']).size().unstack()
plt.figure(figsize=(10, 6))
sns.heatmap(accidents_urban_rural, annot=True, fmt='d', cmap='Blues')
plt.title('Accident Severity by Urban and Rural Areas')
plt.xlabel('Accident Severity')
plt.ylabel('Area Type')
plt.show()

In [ ]:
corr_matrix = accidents[['Number_of_Vehicles', 'Number_of_Casualties']].corr()

In [ ]:
corr_matrix

In [ ]:
plt.figure(figsize=(8, 6))
corr_matrix = accidents[['Number_of_Vehicles', 'Number_of_Casualties']].corr()
sns.heatmap(corr_matrix, annot=True, cmap='coolwarm')
plt.title('Correlation between Number of Vehicles and Casualties')
plt.show()

<h1> INSIGHT 17. road surface condition with the highest average casuaties per accident</h1>

In [ ]:
road = accidents.groupby("Road_Surface_Conditions")["Number_of_Casualties"].mean().idxmax()

In [ ]:
road

In [ ]:
road_avg_casualties = accidents.groupby('Road_Surface_Conditions')['Number_of_Casualties'].mean().sort_values()
plt.figure(figsize=(10, 6))
road_avg_casualties.plot(kind='barh', color='crimson')
plt.title('Average Casualties by Road Surface Condition')
plt.xlabel('Average Casualties')
plt.ylabel('Road Surface Condition')
plt.show()


<h1> INSIGHT 18. accident per month</h1>

In [ ]:
accident_per_month = accidents['Month'].value_counts()

In [ ]:
accident_per_month

In [ ]:
plt.figure(figsize=(10, 6))
month_names = {
    1: 'January', 2: 'February', 3: 'March', 4: 'April', 5: 'May', 6: 'June', 7: 'July' , 8: 'August', 9: 'September', 10:'October', 11: 'November', 12:'December'}
sns.countplot(data=accidents, x='Month', palette='plasma')
plt.title('Accidents per Month')
plt.xlabel('Month')
plt.ylabel('Number of Accidents')
plt.xticks(rotation=45)
plt.show()